File: src/crewlet/api/routes/fleet.py
"""GET /fleet — every node, what it runs, and what it holds.

``/health`` answers about the process that served it, which is the right
answer to "is this node alive" and the wrong one to every question an
operator actually has behind a load balancer: which node is running that
agent, has the config reached everybody, is a seat being served at all.
Refreshing the page hits a different node and tells a different story.

The lease table already knows. Node presence, seat ownership and the
worker duties are all rows in it, written by whoever holds them and
readable by anyone — so this view is a read, not a fan-out of health
probes to peers that may be mid-drain. The per-node config epoch comes
from the control plane's apply status for the same reason.

Its own limit: leases say who *holds* a seat, not what that node is doing
with it right now. In-flight counts stay on ``/health``, which is the
only place that can honestly answer them.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from starlette.requests import Request
from starlette.responses import JSONResponse

from crewlet._logging import get_logger

logger = get_logger("api.routes.fleet")


class FleetUnavailable(Exception):
    """The lease store could not be read at all.

    Distinct from the degraded no-database answer: without a database
    there is genuinely no fleet, and saying so is the truth. A lease
    store that is configured and failing is an outage, and every surface
    has to report it as one rather than as an empty fleet — an empty
    node list reads as "nothing is running", which is the opposite of
    what an unreadable lease table means.
    """


def _database(app: Any) -> Any:
    from crewlet.db.client import Database

    db = getattr(app.state, "database", None)
    return db if isinstance(db, Database) else None


def _seconds_left(expires_at: Any, now: datetime) -> float:
    try:
        return max(0.0, (expires_at - now).total_seconds())
    except Exception:
        return 0.0


async def get_fleet(request: Request) -> JSONResponse:
    """GET /fleet — nodes, their roles and labels, and their seats."""
    try:
        return JSONResponse(await fleet_payload(request.app))
    except FleetUnavailable:
        return JSONResponse({"error": "lease store unavailable"}, status_code=503)


async def fleet_payload(app: Any) -> dict[str, Any]:
    """The fleet, as both surfaces answer it.

    One function behind the REST route and the websocket query, for the
    reason every other read in ``api/queries.py`` shares one: two
    implementations of "what is the fleet" can disagree, and the moment
    they do, which answer an operator gets depends on which transport
    their browser happened to use.

    Degrades rather than fails. Without a database there is no lease
    table and therefore no fleet — the answer is this one node, which is
    exactly what a single-process deployment is, and saying so beats a
    500 that reads like a broken endpoint.
    """
    from crewlet.config import DEFAULT_NODE_ID
    from crewlet.seat.placement import NodeProfile

    db = _database(app)
    this_node = str(getattr(app.state, "node_id", "") or DEFAULT_NODE_ID)

    if db is None:
        return {
            "nodes": [],
            "seats": [],
            "duties": [],
            "unplaceable": [],
            "unmanned_roles": [],
            "this_node": this_node,
            "degraded": "no database configured — leases are process-local",
        }

    from crewlet.db.leases import LeaseError, LeaseStore

    leases = LeaseStore(db)
    now = datetime.now(UTC)
    try:
        node_rows = await leases.list_live("node:")
        seat_rows = await leases.list_live("seat:")
        duty_rows = await leases.list_live("worker:")
    except LeaseError as exc:
        logger.warning("fleet_read_failed", error=str(exc))
        raise FleetUnavailable(str(exc)) from exc

    # owner is an incarnation (``{node_id}:{random}``); the node id is
    # what an operator recognises, and what everything else is keyed on.
    def _node_of(owner: str) -> str:
        return owner.split(":", 1)[0] if owner else ""

    seats: list[dict[str, Any]] = []
    seats_by_node: dict[str, int] = {}
    for lease in seat_rows:
        handle = lease.resource.removeprefix("seat:")
        node = _node_of(lease.owner)
        seats_by_node[node] = seats_by_node.get(node, 0) + 1
        seats.append(
            {
                "handle": handle,
                "node": node,
                "owner": lease.owner,
                "epoch": lease.epoch,
                "expires_in": _seconds_left(lease.expires_at, now),
            }
        )
    seats.sort(key=lambda s: str(s["handle"]))

    applied = await _apply_status(db)
    nodes: list[dict[str, Any]] = []
    for lease in node_rows:
        node_id = lease.resource.removeprefix("node:")
        profile = NodeProfile.from_meta(node_id, lease.meta)
        status = applied.get(node_id, {})
        nodes.append(
            {
                "id": node_id,
                "roles": sorted(str(r) for r in profile.roles),
                "labels": dict(profile.labels),
                "owner": lease.owner,
                "protocol": lease.protocol,
                "seats": seats_by_node.get(node_id, 0),
                "expires_in": _seconds_left(lease.expires_at, now),
                "config_epoch": status.get("epoch", 0),
                "config_status": status.get("status", ""),
                "config_error": status.get("error", ""),
                # When this node last reported an apply outcome. Fetched
                # and then dropped before: a node that stopped reporting
                # is exactly the one an operator is looking for, and
                # without this its stale row is indistinguishable from a
                # node that reported the same epoch a second ago.
                "config_reported_at": status.get("reported_at", ""),
            }
        )
    nodes.sort(key=lambda n: str(n["id"]))

    duties = sorted(
        (
            {
                "duty": lease.resource.removeprefix("worker:"),
                "node": _node_of(lease.owner),
                "expires_in": _seconds_left(lease.expires_at, now),
            }
            for lease in duty_rows
        ),
        key=lambda d: str(d["duty"]),
    )

    return {
        "nodes": nodes,
        "seats": seats,
        "duties": duties,
        "unplaceable": _unplaceable(app, nodes, seats),
        "unmanned_roles": _unmanned_roles(nodes),
        "this_node": this_node,
        # What the fleet is converging ON, so a lagging node reads as
        # "3 epochs behind 41" rather than as a number with nothing to
        # compare it to.
        "target_epoch": await _target_epoch(db),
    }


async def _target_epoch(db: Any) -> int:
    """The newest activation epoch — the one every node is heading for."""
    try:
        from crewlet.db.config_plane import ConfigPlaneStore

        return int(await ConfigPlaneStore(db).target() or 0)
    except Exception:
        logger.exception("fleet_target_epoch_failed")
        return 0


async def _apply_status(db: Any) -> dict[str, dict[str, Any]]:
    """Per-node config apply state, keyed by node id.

    A node whose epoch lags is not an error by itself — every successful
    rollout produces lag — so this reports the numbers and leaves the
    reading to whoever is looking. See ``docs/concepts/control-plane.md``.
    """
    try:
        from crewlet.db.config_plane import ConfigPlaneStore

        rows = await ConfigPlaneStore(db).fleet()
    except Exception:
        logger.exception("fleet_apply_status_failed")
        return {}
    return {
        str(row["node_id"]): {
            "epoch": int(row["epoch"] or 0),
            "status": str(row["status"] or ""),
            "error": str(row["error"] or ""),
            "reported_at": _iso(row.get("updated_at")),
        }
        for row in rows
    }


def _iso(value: Any) -> str:
    try:
        return value.isoformat() if value is not None else ""
    except Exception:
        return ""


def _unmanned_roles(nodes: list[dict[str, Any]]) -> list[str]:
    """Roles no live node performs — the shape no node's config is wrong for."""
    from crewlet.seat.placement import NodeRole

    manned = {role for node in nodes for role in node.get("roles", [])}
    return sorted(str(r) for r in NodeRole if str(r) not in manned)


def _unplaceable(
    app: Any, nodes: list[dict[str, Any]], seats: list[dict[str, Any]]
) -> list[dict[str, str]]:
    """Seats in the org that no live node is eligible to run.

    Computed here rather than reported by a node, because the node that
    reports it is the one that noticed — and when the pinned node is the
    one that died, the survivors are exactly who should be able to say
    so.

    Read off ``agent_roles``, the same projection ``GET /agents`` uses,
    so the handle derivation and the human-seat exclusion are shared. A
    second walk over the payload would have to repeat both, and a seat
    whose handle it derived differently would show here as unplaceable
    forever.
    """
    from crewlet.seat.placement import NodeProfile, SeatPlacement, parse_roles

    profiles = [
        NodeProfile(
            node_id=str(n["id"]),
            roles=parse_roles(list(n.get("roles") or [])),
            labels=dict(n.get("labels") or {}),
        )
        for n in nodes
    ]
    seat_nodes = [p for p in profiles if p.runs_seats]
    held = {str(s["handle"]) for s in seats}

    out: list[dict[str, str]] = []
    for role in getattr(app.state, "agent_roles", None) or []:
        handle = str(role.get("handle") or "")
        raw = role.get("placement") or {}
        if not handle or handle in held or not raw:
            continue
        placement = SeatPlacement(
            node=str(raw.get("node") or ""),
            labels={str(k): str(v) for k, v in (raw.get("labels") or {}).items()},
        )
        if any(placement.matches(p.node_id, dict(p.labels)) for p in seat_nodes):
            continue
        out.append({"handle": handle, "placement": placement.describe()})
    return out

File: internal/api/queries/fleet.go
package queries

import (
	"cmp"
	"context"
	"slices"
	"strings"
	"time"

	"github.com/crewlet/crewlet/internal/coord"
	"github.com/crewlet/crewlet/internal/objstore"
	"github.com/crewlet/crewlet/internal/seat/placement"
)

// The fleet, read from the LEASE TABLE rather than from a fan-out of /health
// probes.
//
// /health answers about the node that served it, so behind a load balancer a
// refresh tells a different story each time. The lease table is the one place
// that knows which node holds what, and every node reads the same rows.

// fleet answers the fleet question.
func (s Sources) fleet(ctx context.Context, _ Params) (any, error) {
	// The three listings through ONE error path. Written out, they were
	// three identical checks of which a test could only ever exercise the
	// first — and three copies of "the lease table IS the fleet, so an
	// unreadable one must not answer an empty company" is three chances
	// for one of them to stop saying it.
	live := map[coord.Class][]coord.Lease{}
	for _, class := range []coord.Class{coord.ClassNode, coord.ClassSeat, coord.ClassWorker} {
		leases, err := s.Coord.ListLive(ctx, class)
		if err != nil {
			return nil, err
		}
		live[class] = leases
	}
	nodes, seats, duties := live[coord.ClassNode], live[coord.ClassSeat], live[coord.ClassWorker]
	now := s.clock()

	held := map[string]int{}
	seatRows := make([]map[string]any, 0, len(seats))
	for _, lease := range seats {
		node := nodeOf(lease.Owner)
		held[node]++
		seatRows = append(seatRows, map[string]any{
			"handle":     nameIn(coord.ClassSeat, lease.Resource),
			"node":       node,
			"owner":      lease.Owner,
			"epoch":      lease.Epoch,
			"expires_in": secondsLeft(lease.ExpiresAt, now),
		})
	}
	slices.SortFunc(seatRows, func(a, b map[string]any) int {
		return cmp.Compare(a["handle"].(string), b["handle"].(string))
	})

	applied := s.applyStatus(ctx)
	target := s.activation(ctx)
	nodeRows := make([]map[string]any, 0, len(nodes))
	for _, lease := range nodes {
		id := nameIn(coord.ClassNode, lease.Resource)
		profile := placement.FromMeta(id, lease.Meta)
		row := map[string]any{
			"id":         id,
			"roles":      profile.Roles.Names(),
			"labels":     profile.Labels,
			"owner":      lease.Owner,
			"protocol":   lease.Protocol,
			"seats":      held[id],
			"expires_in": secondsLeft(lease.ExpiresAt, now),
		}
		// THE SHARE OF THE OBJECT STORE THIS NODE OFFERS, off its own
		// presence — absent on a node that holds none, never a 0 that
		// reads as a data node offering nothing.
		if profile.ObjectWeight > 0 {
			row["object_weight"] = profile.ObjectWeight
		}
		status := applied[id]
		row["config_epoch"] = status.Epoch
		row["config_status"] = status.Status
		row["config_error"] = status.Error
		// WHICH REVISION THIS NODE IS ON, which the epoch does not say.
		//
		// Carried on the apply record for exactly this reason — its own
		// doc: "the fleet view is read while nodes are mid-transition,
		// and a node still on the previous revision is exactly what an
		// operator is looking for." It was written by every node, stored
		// by the plane, read into this loop, and dropped here, so the
		// screen could compare epoch numbers and never name what they
		// stood for.
		row["config_revision_id"] = status.RevisionID
		// WHEN this node last reported. Without it a node that stopped
		// reporting is indistinguishable from one that reported the same
		// epoch a second ago — and the one that stopped is exactly the
		// one an operator is looking for.
		row["config_reported_at"] = isoOrEmpty(status.UpdatedAt)
		// WHAT THAT NODE IS DOING, off its own presence heartbeat.
		// Absent when the node published none (a peer running a build
		// older than the field), and absent is NOT zero: a confident 0
		// would draw an idle row for a process that is simply not saying.
		if live, ok := coord.StatusFromMeta(lease.Meta); ok {
			row["in_flight"] = live.InFlight
			row["draining"] = live.Draining
			row["started_at"] = isoOrEmpty(live.StartedAt)
			if live.Posture != "" {
				row["posture"] = live.Posture
			}
			// HOW FAR THIS NODE'S OWN COPY HAS COME UP, which is a
			// different question from its config epoch: a node can
			// hold the current revision and still be hydrating the
			// state it derives from the log, and only one of those
			// two makes its seats servable.
			//
			// ABSENT RATHER THAN ZERO when the node published none —
			// the meta only carries them when the total is non-zero
			// — for the reason `in_flight` is absent: a confident 0
			// of 0 reads as "ready" for a process that is simply not
			// saying.
			if live.ProjectionsTotal > 0 {
				row["projections_ready"] = live.ProjectionsReady
				row["projections_total"] = live.ProjectionsTotal
			}
		}
		nodeRows = append(nodeRows, row)
	}
	slices.SortFunc(nodeRows, func(a, b map[string]any) int {
		return cmp.Compare(a["id"].(string), b["id"].(string))
	})

	dutyRows := make([]map[string]any, 0, len(duties))
	for _, lease := range duties {
		dutyRows = append(dutyRows, map[string]any{
			"duty":       nameIn(coord.ClassWorker, lease.Resource),
			"node":       nodeOf(lease.Owner),
			"expires_in": secondsLeft(lease.ExpiresAt, now),
		})
	}
	slices.SortFunc(dutyRows, func(a, b map[string]any) int {
		return cmp.Compare(a["duty"].(string), b["duty"].(string))
	})

	out := map[string]any{
		"nodes": nodeRows, "seats": seatRows, "duties": dutyRows,
		"unplaceable":    s.unplaceable(nodeRows, seatRows),
		"unmanned_roles": unmannedRoles(nodeRows),
		"this_node":      s.NodeID,
		// What the fleet is converging ON, so a lagging node reads as
		// "3 epochs behind 41" rather than as a number with nothing to
		// compare it to.
		// ONE READ, TWO FIELDS. `target_epoch` is what every node row is
		// compared against; `activation` is what that number STANDS FOR
		// — which revision, activated when, with the summary whoever
		// activated it wrote. An operator reading "node-2 is two epochs
		// behind" could not see any of the second half.
		//
		// Read once rather than once per field: they are the same
		// pointer, and two reads of it inside one answer can disagree
		// while an activation lands between them — a screen saying
		// "target 42" beside "revision r-41" describes a fleet that
		// never existed.
		"target_epoch": target.epoch,
		"activation":   target.detail,
	}
	if s.Objects != nil {
		out["objects"] = s.objectMap(ctx)
	}
	return out, nil
}

// ObjectMapReader is the stored placement map, read as every node reads it.
type ObjectMapReader interface {
	ObjectMap(ctx context.Context) (coord.ObjectMapRecord, bool, error)
}

// objectMap is the fleet's placement map as the fleet view shows it.
//
// THE STORED MAP, not this node's cached copy, for the reason the whole view
// reads the lease table: every node gives the same answer. And THREE STATES
// NAMED APART, never folded into an empty list — a map the store would not
// give up (`available: false`), a fleet with no map yet (`placed: false`,
// where no upload can land), and one a newer build wrote that this one cannot
// read (`unreadable: true`) — because each sends an operator somewhere
// different, and "no members" reads as the second of them whichever it was.
func (s Sources) objectMap(ctx context.Context) map[string]any {
	rec, found, err := s.Objects.ObjectMap(ctx)
	switch {
	case err != nil:
		return map[string]any{"available": false}
	case !found:
		return map[string]any{"available": true, "placed": false}
	}
	state, err := objstore.DecodeMapState(rec.Value)
	if err != nil {
		return map[string]any{"available": true, "placed": true, "unreadable": true}
	}
	members := make([]map[string]any, 0, len(state.Map.Members))
	for _, m := range state.Map.Members {
		row := map[string]any{"node": m.Node, "weight": m.Weight}
		if since, gone := state.Absent[m.Node]; gone {
			row["absent_since"] = isoOrEmpty(since)
		}
		members = append(members, row)
	}
	// BOTH COUNTS: the replica count asked for, and how many members hold
	// each chunk now — fewer while the fleet has fewer data nodes than it
	// asks for, which is exactly the shortfall an operator is looking for.
	return map[string]any{
		"available": true, "placed": true,
		"epoch": state.Map.Epoch, "replicas": state.Map.Replicas,
		"copies": state.Map.Size(), "members": members,
	}
}

// applyStatus is each node's last config outcome, keyed by node id.
//
// DEGRADES rather than fails: a fleet answer without the config column is
// still the answer to "which node holds what", and refusing the whole view
// because one of its columns is unreadable would blank the screen an operator
// opens when nodes are dying.
func (s Sources) applyStatus(ctx context.Context) map[string]coord.NodeApply {
	if s.Plane == nil {
		return nil
	}
	rows, err := s.Plane.Fleet(ctx)
	if err != nil {
		log.WarnContext(ctx, "fleet_apply_status_failed", "error", err)
		return nil
	}
	out := make(map[string]coord.NodeApply, len(rows))
	for _, row := range rows {
		out[row.NodeID] = row
	}
	return out
}

// activationTarget is the newest activation as this answer renders it.
//
// The epoch is 0 when there is none or it cannot be read, which is safe
// because the client compares against it only when non-zero — an unknown
// target renders as no comparison rather than as "every node is 41 epochs
// behind".
//
// The detail is NIL rather than an object of empty strings, because "nothing
// has ever been activated" and "the pointer could not be read" both leave a
// screen with nothing to render, and an empty object would have it render a
// revision named "" activated at the zero time. The two are told apart in the
// log, which is where an operator looks for a failure, rather than on a screen
// that has no remedy to offer for either.
type activationTarget struct {
	epoch  int64
	detail any
}

func (s Sources) activation(ctx context.Context) activationTarget {
	if s.Plane == nil {
		return activationTarget{}
	}
	target, found, err := s.Plane.Target(ctx)
	switch {
	case err != nil:
		log.WarnContext(ctx, "fleet_activation_unavailable", "error", err)
		return activationTarget{}
	case !found:
		return activationTarget{}
	}
	return activationTarget{epoch: target.Epoch, detail: map[string]any{
		"epoch":       target.Epoch,
		"revision_id": target.RevisionID,
		"at":          isoOrEmpty(target.At),
		"summary":     target.Summary,
	}}
}

// unplaceable are the seats the company declares that no live node may run.
//
// The question a fleet view exists to answer and the one a list of leases
// cannot: a seat pinned to a label no node carries is not "unclaimed yet", it
// is unclaimable, and it stays that way until somebody changes the config or
// starts a node that matches.
func (s Sources) unplaceable(nodes, seats []map[string]any) []map[string]any {
	if s.Company == nil {
		return []map[string]any{}
	}
	company := s.Company()
	if company == nil {
		return []map[string]any{}
	}
	organization, err := company.Organization()
	if err != nil {
		log.Warn("fleet_unplaceable_failed", "error", err)
		return []map[string]any{}
	}
	claimed := make(map[string]bool, len(seats))
	for _, seat := range seats {
		claimed[seat["handle"].(string)] = true
	}
	profiles := make([]placement.NodeProfile, 0, len(nodes))
	for _, node := range nodes {
		roles, err := placement.ParseRoles(node["roles"].([]string))
		if err != nil {
			// A role this build does not know, which is a peer running a
			// newer one. Its own seats are its business; what this answer
			// must not do is conclude that a seat is unplaceable because
			// the node that can run it uses a word we have not learned.
			log.Warn("fleet_unknown_node_role", "node", node["id"], "error", err)
			roles = placement.DefaultRoles()
		}
		profiles = append(profiles, placement.NodeProfile{
			ID: node["id"].(string), Roles: roles,
			Labels: node["labels"].(map[string]string),
		})
	}

	out := []map[string]any{}
	for role := range organization.AllRoles() {
		if !role.IsAgent() {
			continue
		}
		handle := role.Handle()
		if claimed[handle] {
			continue
		}
		if placementFits(role.Placement, profiles) {
			// Unclaimed but placeable: a node could take it, so this is
			// a moment rather than a fault.
			continue
		}
		out = append(out, map[string]any{
			"handle":    handle,
			"placement": role.Placement,
		})
	}
	slices.SortFunc(out, func(a, b map[string]any) int {
		return cmp.Compare(a["handle"].(string), b["handle"].(string))
	})
	return out
}

// placementFits reports whether any live node may run a seat with this
// placement.
//
// A node that does not run seats does not count, however well it matches: the
// placement selector and the node's ROLE are different constraints, and a
// company whose only label-matching node is an ingress-only node has a seat
// nothing will ever claim.
func placementFits(p placement.SeatPlacement, nodes []placement.NodeProfile) bool {
	for _, node := range nodes {
		if node.RunsSeats() && p.Matches(node.ID, node.Labels) {
			return true
		}
	}
	return false
}

// unmannedRoles are the node roles no live node is running.
//
// A company whose workers role is unmanned still answers webhooks and still
// runs turns; what it stops doing is every scheduled and background duty, with
// no error anywhere. That silence is the whole reason this is a field.
func unmannedRoles(nodes []map[string]any) []string {
	running := map[string]bool{}
	for _, node := range nodes {
		for _, role := range node["roles"].([]string) {
			running[role] = true
		}
	}
	out := []string{}
	for role := range placement.DefaultRoles() {
		if !running[string(role)] {
			out = append(out, string(role))
		}
	}
	slices.Sort(out)
	return out
}

// nodeOf reads the node id out of an owner incarnation ({node}:{random}).
//
// The incarnation is what fences a restarted process; the node id is what an
// operator recognises and what everything else is keyed on.
func nodeOf(owner string) string {
	id, _, _ := strings.Cut(owner, ":")
	return id
}

// secondsLeft is how long a lease has, floored at zero.
//
// Negative would be a lease already gone, which the live listing does not
// return — but a clock that moved between the query and this line would
// produce one, and "-3 seconds left" is not something to render.
func secondsLeft(expires, now time.Time) int {
	if expires.IsZero() {
		return 0
	}
	return max(int(expires.Sub(now)/time.Second), 0)
}

func isoOrEmpty(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.UTC().Format(time.RFC3339Nano)
}

// nameIn is a lease's name within its class, for a listing that already asked
// for that class.
//
// Through [coord.Class.Name] rather than a bare TrimPrefix: the prefix is the
// class AND its separator, and trimming the class alone leaves the separator
// on the front of every handle the dashboard renders.
func nameIn(class coord.Class, resource string) string {
	name, _ := class.Name(resource)
	return name
}
